src/front/glsl/builtins.rs

```diff
@@ -1429,88 +1393,127 @@ impl MacroCall {
+                            op: BinaryOperator::Divide,
+                            left,
+                            right,
+                        },
                         SourceMetadata::none(),
                         body,
                     );
-                    ctx.vector_resize(size, base, SourceMetadata::none(), body)
+                }
+
+                let extra = args.get(2).copied();
+                let comps = parser.coordinate_components(ctx, args[0], coords, extra, meta, body)?;
+
+                let mut num_args = 2;
+
+                if comps.used_extra {
+                    num_args += 1;
                 };
-                let coords = ctx.add_expression(
-                    Expression::Binary {
-                        op: BinaryOperator::Divide,
-                        left,
-                        right,
-                    },
-                    SourceMetadata::none(),
-                    body,
-                );
-                let comps = parser.coordinate_components(ctx, args[0], coords, meta, body)?;
-                texture_call(ctx, args[0], level, comps, body, meta)
+
+                // Parse out explicit texture level.
+                let mut level = match level_type {
+                    TextureLevelType::None => SampleLevel::Auto,
+
+                    TextureLevelType::Lod => {
+                        num_args += 1;
+
+                        if shadow {
+                            log::debug!(
+                                "Assuming LOD {:?} is zero",
+                                args[2],
+                            );
+
+                            SampleLevel::Zero
+                        } else {
+                            SampleLevel::Exact(args[2])
+                        }
+                    }
+
+                    TextureLevelType::Grad => {
+                        num_args += 2;
+
+                        if shadow {
+                            log::debug!(
+                                "Assuming gradients {:?} and {:?} are not greater than 1",
+                                args[2],
+                                args[3],
+                            );
+                            SampleLevel::Zero
+                        } else {
+                            SampleLevel::Gradient {
+                                x: args[2],
+                                y: args[3],
+                            }
+                        }
+                    }
+                };
+
+                let texture_offset = match offset {
+                    true => {
+                        let offset_arg = args[num_args];
+                        num_args += 1;
+                        match parser.solve_constant(ctx, offset_arg, meta) {
+                            Ok(v) => Some(v),
+                            Err(e) => {
+                                parser.errors.push(e);
+                                None
```

Comment:
Very nice, non exiting error reporting is always good to have when it doesn't break further error handling

Reply:
It would be nice to have a helper for this eventually.